### ISSUE TITLE
Preformence: change TopKMaskLogitsKernel to oveewrite the tensor instead of allocating a new tensor

### ISSUE THREAD

**Issue by yonikremer** [flashinfer-ai/flashinfer#1900]
Hi, I am tying to scale up to large batch sized and noticed that TopKMaskLogitsKernel is allocating a new tensor instead of modified the tensor in-place.
In the vast majority of cases you are not interster in the orginial logits, but very intrested in saving GPU memory.
looking at the code of the kernel, I can see clearlly we write to the output onlyu after we finish reading from the input.
I am willing to try to contribute this if you want

**Comment by yzh119** [comment]
Yes, this would be a great feature to have! Thank you for the suggestion.
I would encourage adding an optional `out=` argument to these APIs. The behavior could work as follows:

* If out is specified, the result would be written to the user-specified tensor
* If out is not specified, a new tensor would be allocated and the result written to it

This design would naturally support the inplace operations you mentioned in your issue: users could simply pass the same tensor as both the input and the out argument to achieve inplace behavior. At kernel-wise, there is nothing to change.

Reference can be found at https://github.com/flashinfer-ai/flashinfer/blob/c3ff7e76e67d7ca6875ed6137322eedf5c356f32/flashinfer/rope.py, see how inplace operations are implemented.

**Comment by yonikremer** [comment]
I started thinking about it, in most cases, top k is very small comared to the vocab size (100 vs 100k), maybe storing the results as a sparse tensor would be a better idea?

**Comment by yonikremer** [comment]
I think that later on computing sotmax and sampling from a sparse tesnor should be much much faster

**Comment by yzh119** [comment]
> maybe storing the results as a sparse tensor would be a better idea?

Yes that's something I have been brainstorming w/ @xslingcn before but we had no bandwidth working on this.
